### PR TITLE
Predefined script, add option for user to download kubeconfig

### PR DESCRIPTION
**Pull request by tuannvm**
## Why?
- predefined script, add option for user to download kubeconfig

## Screenshot
image

### CODE REVIEW

File: kubeconfig.tpl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: REDACTED
```
Comment (so0k) [inline kubeconfig.tpl]: Hmmm REDACTED, should be placeholder and part of dex-app config?

File: kubeconfig.tpl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    server: https://api.ap-southeast-1a.staging.k8s.honestbee.com
+  name: ap-southeast-1a.staging.k8s.honestbee.com
```
Comment (so0k) [inline kubeconfig.tpl]: Should be configurable?

File: kubeconfig.tpl [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+current-context: k8s-1.7-staging
+kind: Config
+preferences: {}
+users:
+- name: k8s-user
+  user:
+    auth-provider:
+      config:
+        client-id: kubernetes
+        client-secret: AtMUIzMy00ODg0LTkwMDQtME
```
Comment (so0k) [inline kubeconfig.tpl]: Render this in, part of config already

File: kubeconfig.tpl [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        extra-scopes: groups
+        id-token: {{ .IDToken }}
+        idp-issuer-url: https://dex.honestbee.com
```
Comment (so0k) [inline kubeconfig.tpl]: This also

File: templates.go [hunk 5]
```diff
@@ -47,20 +44,44 @@ pre {
...
+		<input type="submit" value="Download Kubeconfig">
+	</form>
   </body>
 </html>
 `))
 
+var kubeConfigTmpl = template.Must(template.New("kubeconfig.tpl").ParseFiles("kubeconfig.tpl"))
+
+func renderIndex(w http.ResponseWriter) {
+	renderTemplate(w, indexTmpl, nil)
```
Comment (so0k) [inline templates.go]: Cool

File: templates.go [hunk 6]
```diff
@@ -70,6 +91,15 @@ func renderToken(w http.ResponseWriter, redirectURL, idToken, refreshToken strin
 	})
 }
 
+func renderKubeConfig(w http.ResponseWriter, idToken, refreshToken string) {
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", "kubeconfig"))
+	renderTemplate(w, kubeConfigTmpl, tokenTmplData{
```
Comment (so0k) [inline templates.go]: Pass in dex-app config also?

File: kubeconfig.tpl (hunk 1 above)
Comment (tuannvm) [reply kubeconfig.tpl]: Not sure what you mean? So we should include the CA data of each clusters in dex-app config, and render it?

File: kubeconfig.tpl (hunk 2 above)
Comment (tuannvm) [reply kubeconfig.tpl]: yea

File: kubeconfig.tpl (hunk 3 above)
Comment (tuannvm) [reply kubeconfig.tpl]: ok

File: templates.go (hunk 6 above)
Comment (tuannvm) [reply templates.go]: ok

File: main.go [hunk 7]
```diff
@@ -250,10 +256,25 @@ func (a *app) handleLogin(w http.ResponseWriter, r *http.Request) {
 		authCodeURL = a.oauth2Config(scopes).AuthCodeURL(exampleAppState, oauth2.AccessTypeOffline)
 	}
 
+	ClientID = clients[0]
 	http.Redirect(w, r, authCodeURL, http.StatusSeeOther)
 }
 
 func (a *app) handleDownload(w http.ResponseWriter, r *http.Request) {
+	m := map[string]map[string]string{
```
Comment (tuannvm) [inline main.go]: predefined map and get value from `env`

File: main.go [hunk 8]
```diff
@@ -266,7 +287,7 @@ func (a *app) handleDownload(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	renderKubeConfig(w, refreshToken, idToken)
+	renderKubeConfig(w, clientID, caCert, clusterEndpoint, refreshToken, idToken)
```
Comment (tuannvm) [inline main.go]: render `kubeconfig` with new values

File: templates.go [hunk 9]
```diff
@@ -8,75 +8,20 @@ import (
...
+	CACert          string
+	ClientID        string
+	ClusterEndpoint string
+	IDToken         string
+	RefreshToken    string
+	RedirectURL     string
+	Claims          string
 }
 
-var indexTmpl = template.Must(template.New("index.html").Parse(`<html>
```
Comment (tuannvm) [inline templates.go]: split template out of the source codes

**Comment by tuannvm** [comment]
@so0k Please take a look, thanks.

File: main.go [hunk 10]
```diff
@@ -23,7 +23,10 @@ import (
 	"golang.org/x/oauth2"
 )
 
-const exampleAppState = "I wish to wash my irish wristwatch"
+const exampleAppState = "Honestbee FTW"
+
+// ClientID : global variable to take client_id and pass to /download
+var ClientID = ""
```
Comment (tuannvm) [inline main.go]: global variable to get `ClientID` from `/login`

File: main.go [hunk 11]
```diff
@@ -23,7 +23,22 @@ import (
...
+
+// M : predefined map to help lookup values
```
Comment (so0k) [inline main.go]: variable name is M 😱

Comment (so0k) [reply main.go]: it's a map for cluster definitions (endpoint and cacert) - variable name should reflect that purpose

File: main.go [hunk 12]
```diff
@@ -192,8 +211,8 @@ func cmd() *cobra.Command {
 			}
 		},
 	}
-	c.Flags().StringVar(&a.clientID, "client-id", "example-app", "OAuth2 client ID of this application.")
-	c.Flags().StringVar(&a.clientSecret, "client-secret", "ZXhhbXBsZS1hcHAtc2VjcmV0", "OAuth2 client secret of this application.")
+	c.Flags().StringVar(&a.clientID, "client-id", "", "OAuth2 client ID of this application.")
+	c.Flags().StringVar(&a.clientSecret, "client-secret", "", "OAuth2 client secret of this application.")
```
Comment (so0k) [inline main.go]: I wouldn't remove the defaults, what's the harm in having defaults?

We just make sure we don't use the defaults when we deploy the app..

Comment (so0k) [reply main.go]: seems [Cobra](https://github.com/spf13/cobra) only manages flags and doesn't work with env vars.

For that we'd need to factor in [Viper](https://github.com/spf13/viper) to handle config more appropriately

I'm more used to [urfave/cli](https://github.com/urfave/cli) - which allows flags to be set from Env vars

File: main.go [hunk 13]
```diff
@@ -229,7 +248,10 @@ func (a *app) handleLogin(w http.ResponseWriter, r *http.Request) {
 	var scopes []string
 	if extraScopes := r.FormValue("extra_scopes"); extraScopes != "" {
 		scopes = strings.Split(extraScopes, " ")
+	} else {
+		scopes = append(scopes, os.Getenv("SCOPES"))
```
Comment (so0k) [inline main.go]: config is sprinkled throughout - I think it's fine for now, but I should consider adding Viper support to manage config better.

File: main.go [hunk 14]
```diff
@@ -178,8 +193,12 @@ func cmd() *cobra.Command {
 
 			http.HandleFunc("/", a.handleIndex)
 			http.HandleFunc("/login", a.handleLogin)
+			http.HandleFunc("/download", a.handleDownload)
 			http.HandleFunc(u.Path, a.handleCallback)
 
+			fs := http.FileServer(http.Dir("assets"))
```
Comment (so0k) [inline main.go]: interesting

File: main.go (hunk 11 above)
Comment (so0k) [reply main.go]: those are client clusters and the index is `clientID` ... so maybe it should be called `clientClusters` ?

File: main.go [hunk 15]
```diff
@@ -249,9 +271,26 @@ func (a *app) handleLogin(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, fmt.Sprintf("no refresh_token in request: %q", r.Form), http.StatusBadRequest)
+		return
+	}
+
+	if idToken == "" {
+		http.Error(w, fmt.Sprintf("no id_token in request: %q", r.Form), http.StatusBadRequest)
+		return
+	}
+
+	renderKubeConfig(w, ClientID, M[ClientID]["CACert"], M[ClientID]["ClusterEndpoint"], refreshToken, idToken)
```
Comment (so0k) [inline main.go]: we'd have `clientClusters[ClientID]["CACert"]` which is very clear opposed to `M`

File: templates/index.html [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+      <link rel="stylesheet" media="all" href="/assets/application-859466a8b944ec07c87053d1af5e7794612ca74f13d686e00d4c8410b24ab02b.css" />
+      <script src="/assets/application-5c1692b6030dd1b338cf8aaed8f4ae2c1b938a0130620a2b0e8e540500187d57.js"></script>
```
Comment (so0k) [inline templates/index.html]: where is this js file? forgot to commit or am I being daft?

**Comment by so0k** [comment]
I will summarize what I understand you did:

- Scopes are now managed through env var when app is started instead of passed in form
- page templates have been properly moved to a templates directory
- the focus is on k8s clusters as client applications, their configuration has been moved to a map which is currently hardcoded to read 2 env variables.
- other config is managed by reading env vars directly

I only have 1 comment on the variable naming... but I think it's good enough to get things done for now - can always improve later

File: templates/index.html (hunk 16 above)
Comment (tuannvm) [reply templates/index.html]: Oh snap, yea forgot to commit :(

File: main.go (hunk 11 above)
Comment (tuannvm) [reply main.go]: 🤔 possibly. Because I use the existing naming convention for `clientID`

File: main.go (hunk 12 above)
Comment (tuannvm) [reply main.go]: Thanks, good to know.

File: main.go (hunk 15 above)
Comment (tuannvm) [reply main.go]: Right :D

File: main.go (hunk 13 above)
Comment (tuannvm) [reply main.go]: Will look onto that.

**Comment by tuannvm** [comment]
### Some notes for improvements:
* Looks into [Viper](https://github.com/spf13/viper) and [urfave/cli](https://github.com/urfave/cli) to improve env support
* Consider to replace global variables

File: templates/kubeconfig.tpl [hunk 17]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: REDACTED
```
Comment (so0k) [inline templates/kubeconfig.tpl]: oops, is going to render as expected?

Comment (tuannvm) [reply templates/kubeconfig.tpl]: oh snap, hot fix incoming...

**Comment by so0k** [comment]
just creating cross repo pr references

this supports: https://github.com/honestbee/helm-charts/pull/157

**Pull request merged**